### ISSUE TITLE
Emmet is not working as expected

### ISSUE THREAD

**Issue by tobias-kuendig** [tomblachut/svelte-intellij#19]
Can this be because of this plugin? I'm using Goland.

I haven't found out when exaclty it doesn't work. It never works inside another tag. Sometimes it doesn't even work on the root level.

Changing the File type of `.svelte` files back to HTML makes everything work as expected again.

![emmet](https://user-images.githubusercontent.com/8600029/58103408-334d3000-7be3-11e9-8adb-50f7178a3790.gif)

**Comment by tomblachut** [comment]
Really weird, thanks for raising!

**Comment by unlocomqx** [comment]
Here's some debug info

In 
`com.intellij.codeInsight.template.emmet.XmlEmmetParser#getDefaultTemplateKey`
then 
`com.intellij.codeInsight.template.emmet.ZenCodingUtil#isHtml` ,
the language is checked and must be HTML or XML and so on
To pass that condition, you need to override `getBaseLanguage` in `SvelteLanguage` and return `HtmlLanguage.INSTANCE`

Even so, emmet fails because it creates the wrong dummy file type (SvelteFile instead of HTML File)
This is because it reads the file language (which is SvelteLanguage) and doesn't look for a base language that is HTML
So then `com.intellij.codeInsight.template.emmet.tokens.TemplateToken#getXmlTag` fails to find the XML tag that was added dynamically and null is returned
![screenshot_16](https://user-images.githubusercontent.com/793712/61241059-a1712780-a73a-11e9-9986-a61e097d5284.png)

So basically `com.intellij.codeInsight.template.emmet.tokens.TemplateToken#setTemplateText` should use the HTML language instead of the file language which it gets through `callback.getFile().getLanguage()`

Or we should find a way to allow `PsiTreeUtil` to find an `XmlTag` inside `HTML_FRAGMENT` elements

**Comment by tomblachut** [comment]
@unlocomqx thx to your finding about `setTemplateText` I managed to do *something*
For example
```
    <ul>
        {#each assets as {title} (a.id)}
            .header
        {/each}
    </ul>
```
Is expanded into
```
    <ul>
        {#each assets as {title} (a.id)}
            <li class=""></li>
        {/each}
    </ul>
```

So finding a parent works even across Svelte outer fragments which is great. Expanding if blocks etc. still works. Figuring out how to fix complex emmet patterns will be easy enough I hope.

My approach revolves around updating `SvelteZenCodingGeneratorImpl`

```
    override fun createParser(tokens: MutableList<ZenCodingToken>?, callback: CustomTemplateCallback, generator: ZenCodingGenerator?, surroundWithTemplate: Boolean): EmmetParser {
        return XmlEmmetParser(tokens, SvelteZenCodingCustomTemplateCallback(callback), generator, surroundWithTemplate)
    }
```

```
class SvelteZenCodingCustomTemplateCallback(callback: CustomTemplateCallback) : CustomTemplateCallback(callback.editor, callback.file.viewProvider.getPsi(HTMLLanguage.INSTANCE))
```
Notice that I overrode PsiFile in callback to always be HTML one.

**Comment by unlocomqx** [comment]
Brilliant
I will try to figure out why it doesn't assign the attribute value correctly

**Comment by unlocomqx** [comment]
I did some debugging and found that this line gets the Psi File with Svelte Language when cursor is inside a Svelte Fragment

TemplateManagerImpl.java:253
`PsiFile file = PsiUtilBase.getPsiFileInEditor(editor, myProject);`

That file is used to instantiate CustomTemplateCallback in
TemplateManagerImpl.java:263
`CustomTemplateCallback templateCallback = new CustomTemplateCallback(editor, file);`

(is there a way to override that?)

Then a dummy file is created using Svelte Language here
GenerationNode.java:274
`PsiFile dummyFile = fileFactory.createFileFromText(...`

In the next line PsiTreeUtil.findChildOfType can't find the tag and null is returned which prevents entering the if which contains a call to `setAttributeValues`

Cursor inside SVELTE_FRAGMENT
![image](https://user-images.githubusercontent.com/793712/61631452-5008f100-ac82-11e9-8671-769514d1aa66.png)

Result
![image](https://user-images.githubusercontent.com/793712/61631481-631bc100-ac82-11e9-95e6-fef05199a1ff.png)

**Comment by tomblachut** [comment]
In #40 I've changed how whitespace is parsed.

Now it's delegated to HTML side with notable exception of continuous whitespace between Svelte mustaches and first non-whitespace HTML character, it's required for current (bad) implementation of formatter to work.

Anyway that fixed Emmet partially.

![image](https://user-images.githubusercontent.com/615013/62013758-65f14700-b197-11e9-8a41-e1975a0bda49.png)

![image](https://user-images.githubusercontent.com/615013/62013983-4a883b00-b19b-11e9-86b9-f1c0b19ed573.png)


Preview still does not work with the last character in a file. Add one space after Emmet shorthand and everything starts to work. Anyway you can expand shorthand and it works in both cases.

![image](https://user-images.githubusercontent.com/615013/62013809-43136280-b198-11e9-8658-22f07b568e42.png)

Remaining unsolved issue is shorthand directly after Svelte mustache. I suspect this is connected to mentioned whitespace handling and by extension formatter 
![image](https://user-images.githubusercontent.com/615013/62013951-fd0bce00-b19a-11e9-8485-777917e69e98.png)

**Comment by unlocomqx** [comment]
That's great
IMO, the edge cases are not severe 
This one is because the document ends with a SVELTE_FRAGMENT
![image](https://user-images.githubusercontent.com/793712/62034965-d3789400-b1e6-11e9-844d-1be92010e1eb.png)

![image](https://user-images.githubusercontent.com/793712/62037224-4dab1780-b1eb-11e9-961d-6d54a78a204d.png)

**Comment by tomblachut** [comment]
@unlocomqx in branch with lexer fixes document actually ends with XmlToken

**Comment by unlocomqx** [comment]
I just tested with rework-lexer, it worked correctly
![image](https://user-images.githubusercontent.com/793712/62083189-49641600-b24e-11e9-80e9-53e2be50622e.png)

**Comment by tomblachut** [comment]
yeah it expands correctly but preview does not work for me as shown in screenshot

**Comment by unlocomqx** [comment]
Sry for the confusion, how do you get the preview to pop up? is it automatic?

**Comment by tomblachut** [comment]
No worries. You need to enable it and then it's automatic

![image](https://user-images.githubusercontent.com/615013/62083532-6d782500-b257-11e9-90c6-a50eda3246b0.png)

**Comment by unlocomqx** [comment]
I was able to reproduce,
It's not a big deal really but I will try to debug it sometime nonetheless

**Comment by unlocomqx** [comment]
As I expected, it's the same as this

> Then a dummy file is created using Svelte Language here
> GenerationNode.java:274
> `PsiFile dummyFile = fileFactory.createFileFromText(...`
> 
> In the next line `PsiTreeUtil.findChildOfType` can't find the tag and null is returned which prevents entering the if which contains a call to `setAttributeValues`
